cirq/protocols/json_serialization.py

```diff
@@ -429,6 +446,128 @@ def _cirq_object_hook(d, resolvers: Sequence[JsonResolver]):
+        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
+
+    @classmethod
+    def read_from_context(cls, context_map, key, **kwargs):
+        return context_map[key]
+
+
+class _SerializedContext(SupportsJSON):
+    """Internal object for a single SerializableByKey key-to-object mapping.
+
+    This is a private type used in contextual serialization. Its deserialization
+    is context-dependent, and is not expected to match the original; in other
+    words, `cls._from_json_dict_(obj._json_dict_())` does not return
+    the original `obj` for this type.
+    """
+
+    def __init__(self, obj: SerializableByKey):
+        self.key = obj._serialization_key_()
+        self.obj = obj
+
+    def _json_dict_(self):
+        return obj_to_dict_helper(self, ['key', 'obj'])
+
+    @classmethod
+    def _from_json_dict_(cls, **kwargs):
+        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
+
+    @classmethod
+    def update_context(cls, context_map, key, obj, **kwargs):
+        context_map.update({key: obj})
+
+
+class _ContextualSerialization(SupportsJSON):
+    """Internal object for serializing an object with its context.
+
+    This is a private type used in contextual serialization. Its deserialization
+    is context-dependent, and is not expected to match the original; in other
+    words, `cls._from_json_dict_(obj._json_dict_())` does not return
+    the original `obj` for this type.
+    """
+
+    def __init__(self, obj: Any):
+        self.context_list = []
+        context_keys = set()
+        for sbk in get_serializable_by_keys(obj):
+            new_sc = _SerializedContext(sbk)
+            if new_sc.key not in context_keys:
+                self.context_list.append(new_sc)
+                context_keys.add(new_sc.key)
+        self.context_list += [obj]
+
+    def _json_dict_(self):
+        return obj_to_dict_helper(self, ['context_list'])
+
+    @classmethod
+    def _from_json_dict_(cls, **kwargs):
+        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
+
+    @classmethod
+    def deserialize_with_context(cls, context_list, **kwargs):
+        return context_list[-1]
+
+
+def has_serializable_by_keys(obj: Any) -> bool:
+    """Returns true if obj contains one or more SerializableByKey objects."""
+    if hasattr(obj, '_serialization_key_'):
+        return True
+    json_dict = getattr(obj, '_json_dict_', lambda: None)()
+    if isinstance(json_dict, Dict):
+        return any(has_serializable_by_keys(v) for v in json_dict.values())
+    return False
+
+
+def get_serializable_by_keys(obj: Any) -> List[SerializableByKey]:
+    """Returns all SerializableByKeys contained by obj.
+
+    Objects are ordered such that nested objects appear before the object they
+    are nested inside. This is required for hooks in json.load().
+    """
+    result = []
+    if hasattr(obj, '_serialization_key_'):
```

Comment:
While this works as a single round of simplification, and especially when SBK objects are nested under non-SBK objects. However, it won't really work when SBK objects are nested? Not sure how to work this out, or how frequent this will be but that case the context will contain the potential structural repetitions - e.g.:


<details>

```
{
  "cirq_type": "_ContextualSerialization",
  "context_list": [
    {
      "cirq_type": "_SerializedContext",
      "key": "a",
      "value": {
        "cirq_type": "CircuitOperation",
        "ops": [
          {"cirq_type":  "X"},
          {
            "cirq_type": "CircuitOperation",
            "ops": ["big list of operations"]
          }
        ]
      }
    },
    {
      "cirq_type": "_SerializedContext",
      "key": "b",
      "value": {
        "cirq_type": "CircuitOperation",
        "ops": [
          {"cirq_type":  "Y"},
          {
            "cirq_type": "CircuitOperation",
            "ops": ["big list of operations"]
          }
        ]
      }
    },
    {
      "cirq_type": "_SerializedContext",
      "key": "c",
      "value": {
        "cirq_type": "CircuitOperation",
        "ops": [
          {"cirq_type":  "Z"},
          {
            "cirq_type": "CircuitOperation",
            "ops": ["big list of operations"]
          }
        ]
      }
    },
    {
      "cirq_type": "Circuit",
      "moments": [
        {
          "cirq_type": "GateOperation",
          "gate": {
            "cirq_type": "_SerializedKey",
            "key": "a",
          },
          "qubits": a
        },
        {
          "cirq_type": "GateOperation",
          "gate": {
            "cirq_type": "_SerializedKey",
            "key": "b",
          },
          "qubits": b
        },
        {
          "cirq_type": "GateOperation",
          "gate": {
            "cirq_type": "_SerializedKey",
            "key": "c",
          },
          "qubits": c
        }
      ]
    }
  ]
}
```

</details>

Reply:
Note that we do not immediately return `obj`. If the SBK has a `_json_dict_` method - and it must, since SBK is a `SupportsJSON` type - then all nested SBKs will be populated into the context as well.

This behavior is tested in `test_context_serialization`.

Reply:
Got it - I did misread result.append as return lol :D